File: leetcode/19.cpp
// Slow and fast pointer.

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (!head)
            return NULL;
        
        ListNode *slow = head, *fast = head;
        for (int i = 0; i < n; i++) {
            fast = fast->next;
        }
        
        if (!fast)
            return head->next;
        
        while (fast->next) {
            slow = slow->next;
            fast = fast->next;
        }
        
        slow->next = slow->next->next;
        return head;
    }
};

File: leetcode/19.rs
impl Solution {
    fn help(head: Option<Box<ListNode>>, n: i32) -> (Option<Box<ListNode>>, i32) {
        match head {
            None => {
                return (None, 0);
            }
            Some(box_) => {
                let v = box_.val;
                let (nxt, i) = Self::help(box_.next, n);
                if i + 1 == n {
                    return (nxt, i + 1);
                } else {
                    let newn = Some(Box::new(
                        ListNode {
                            val: v,
                            next: nxt,
                        }
                    ));
                    return (newn, i + 1);
                }
            }
        }
    }

    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let (x, _) = Self::help(head, n);
        return x;
    }
}
